src/lightning/fabric/plugins/precision/fp8_transformer_engine.py

```diff
@@ -0,0 +1,116 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import logging
+from contextlib import contextmanager
+from typing import Any, Dict, Generator, Literal, Optional, TYPE_CHECKING, Union
+
+import torch
+from lightning_utilities.core.imports import RequirementCache
+
+from lightning.fabric.plugins.precision.precision import Precision
+from lightning.fabric.utilities.rank_zero import rank_zero_warn
+
+_TRANSFORMER_ENGINE_AVAILABLE = RequirementCache("transformer_engine")
+
+if TYPE_CHECKING and _TRANSFORMER_ENGINE_AVAILABLE:
+    from transformer_engine.common.recipe import DelayedScaling
+else:
+    DelayedScaling = None
+
+
+log = logging.getLogger(__name__)
+
+
+class Fp8TransformerEnginePrecision(Precision):
+    """Plugin for training with fp8 precision via nvidia's `Transformer Engine
+    <https://docs.nvidia.com/deeplearning/transformer-engine`__.
+
+    .. warning::  This is an :ref:`experimental <versioning:Experimental API>` feature.
+
+    Args:
+        recipe: Recipe for the DelayedScaling
+            `configuration <https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/common.html#transform
+            er_engine.common.recipe.DelayedScaling`__. In dict format or the dataclass format.
+        replace_layers: Whether to replace ``Linear`` and ``LayerNorm`` layers automatically with their Transformer
+            Engine alternatives.
+    """
+
+    precision: Literal["8-mixed-transformer-engine"] = "8-mixed-transformer-engine"
+
+    def __init__(
+        self, recipe: Optional[Union[Dict[str, Any], "DelayedScaling"]] = None, replace_layers: bool = True
+    ) -> None:
+        if not torch.cuda.get_device_capability() >= (8, 1):
+            raise NotImplementedError("Your CUDA device does not support fp8 mixed precision.")
+        if not _TRANSFORMER_ENGINE_AVAILABLE:
+            raise ModuleNotFoundError(str(_TRANSFORMER_ENGINE_AVAILABLE))
+        from transformer_engine.common.recipe import DelayedScaling
+
+        if recipe is None:
+            recipe = DelayedScaling()
+        elif isinstance(recipe, dict):
+            if "fp8_format" in recipe:
+                from transformer_engine.common.recipe import Format
+
+                recipe["fp8_format"] = getattr(Format, recipe["fp8_format"])
+            recipe = DelayedScaling(**recipe)
+
+        self.recipe = recipe
+        self.replace_layers = replace_layers
+
+    @contextmanager
+    def forward_context(self) -> Generator[None, None, None]:
+        import transformer_engine.pytorch as te
+
+        with te.fp8_autocast(enabled=True, fp8_recipe=self.recipe):
+            yield
+
+    def convert_module(self, module: torch.nn.Module) -> torch.nn.Module:
```

Comment:
This will work but could you also implement the variant in `module_init_context` which will be more efficient than converting after the fact.

Reply:
Are you suggesting that in addition to this?

I went for `convert_module` as `init_module` is optional. Your suggestion is that we try to do it if it's used and here otherwise?

Reply:
Yes that's what I meant. You already have the logic in convert_module to prevent doing it if layers are already of this type.

Doing it as part of init_module() is the most efficient IMO, as you only allocate the memory that is needed. Applying the layers directly at the time of construction of the module means we can fit models on a GPU that would otherwise not fit if at first the full fp32 memory gets allocated.

It would of course require patching and this is less flexible, and that means we still need to verify that the idea as a whole is practical.